framework/db/Query.php

```diff
@@ -590,12 +590,47 @@ public function addSelect($columns)
         if ($this->select === null) {
             $this->select = $columns;
         } else {
-            $this->select = array_merge($this->select, $columns);
+            $this->select = array_merge($this->select, $this->removeDuplicatedColumns($columns));
         }
 
         return $this;
     }
 
+    /**
+     * Removes duplicated columns from given columns definitions.
+     * Columns to be removed:
+     * - if column definition already present in SELECT part with same alias
+     * - if column definition without alias already present in SELECT part without alias too
+     * @param array $columns the columns to be merged to the select.
+     */
+    protected function removeDuplicatedColumns($columns)
+    {
+        $simple_columns = $this->filterSimpleColumnsFromSelect();
+
+        foreach ($columns as $columnName => $columnDefinition) {
+            if ($columnDefinition instanceof Query)
+                continue;
+
+            if ((is_string($columnName) && isset($this->select[$columnName]) && $this->select[$columnName] === $columnDefinition) ||
+                (is_integer($columnName) && in_array($columnDefinition, $simple_columns)))
+                unset($columns[$columnName]);
+        }
+        return $columns;
+    }
+
+    /**
+     * @return array List of simple columns from query SELECT
+     */
+    protected function filterSimpleColumnsFromSelect()
+    {
+        $result = [];
+        foreach ($this->select as $name => $value) {
+            if (is_integer($name) && !in_array($value, $result))
```

Comment:
Have you checked if not doing `in_array` here and instead calling `array_unique` right before return will be better for performance?

Reply:
Missing `{` and `}`.